### PR TITLE
[batch]Add back button

### PR DESCRIPTION
**Pull request by CDiaz96**

### CODE REVIEW

**Comment by danking** [comment]
I'm dev deploying now to take a look!

**Review by danking — CHANGES_REQUESTED** [review]
This is harder than I thought it was!

Let us start from scratch. When I ask for a page, I will always ask for a particular batch id and a
direction. For example, if the newest batch id is 14, then "10, previous" might look like:

| id | name | cost | ... |
| -- | ---- | ---- | --- |
| 14 | foo  | 0.01 | ... |
| 13 | bar  | 0.04 | ... |
| 12 | baz  | 0.02 | ... |
| 11 | qux  | 0.03 | ... |
| 10 | qux  | 0.03 | ... |

In contrast, "10, next" might look like:

| id | name | cost | ... |
| -- | ---- | ---- | --- |
| 10 | foo  | 0.01 | ... |
|  9 | foo  | 0.01 | ... |
|  8 | bar  | 0.04 | ... |
|  7 | baz  | 0.02 | ... |
|  6 | qux  | 0.03 | ... |
|  5 | foo  | 0.01 | ... |
|  4 | bar  | 0.04 | ... |
|  3 | baz  | 0.02 | ... |
|  2 | qux  | 0.03 | ... |
|  1 | foo  | 0.01 | ... |

This is almost what we do now with `last_batch_id` and `first_batch_id`. However, the more I changed
what we do to make it correct, the more those names felt wrong. I will eventually propose we instead
use `next_page_first_batch_id` and `previous_page_last_batch_id`.

Before I propose that, let's look at what we are currently doing:

```sql
select id
from batches
where id < last_batch_id
order by batches.id
limit 50
```

And then we have

```python3
if len(batches) == 50:
    last_batch_id = batches[-1]['id']
else:
    last_batch_id = None
```

Which is actually wrong! If we have exactly 50 batches in the entire database, then we're at the
last (and first!) page and we should set `last_batch_id` to `None` (which means we show no "next
page" button). That's OK, we can fix that easily. We'll ask for 51 batches. If we see 51 batches,
then we know there's at least one more page to show. We show the user the first 50 batches and a
"next page" button.

```sql
select id
from batches
where id < last_batch_id
order by batches.id DESC
limit 51
```

```python3
if len(batches) == 51:
    last_batch_id = batches[-1]['id']  # this means we show a next page button
else:
    last_batch_id = None

batches = batches[:50]  # only show the user at most 50 batches
```

But how do we know if there is a previous page? There is a previous page if there is at least one
batch with an id greater than or equal to `last_batch_id`. My simplest idea is to issue a second SQL
query:

```sql
select id
from batches
where id >= last_batch_id
limit 1
```

If that query returns one record, then there is a previous page. If that query returns no records,
there is no previous page.

---

OK, now my proposal.

I think we should replace `last_batch_id` with `next_page_first_batch_id`. Instead of telling the
server what the last batch id of the current page is, it tells the server we need the next page and
that page should start with the given batch id. We should replace `first_batch_id` with
`previous_page_last_batch_id`. It tells the server we need the previous page and that page should
end with the given batch id.

When you first go to `/batches`, both of these are `None`.

Inside `_query_batches`, this block:

```python3
    last_batch_id = request.query.get('last_batch_id')
    if last_batch_id is not None:
        last_batch_id = int(last_batch_id)
        where_conditions.append('(id < %s)')
        where_args.append(last_batch_id)
```

changes to:

```python3
    previous_page_last_batch_id = request.query.get('previous_page_last_batch_id')
    next_page_first_batch_id = request.query.get('next_page_first_batch_id')

    if previous_page_last_batch_id is not None:
        assert next_page_first_batch_id is None

        order_by_direction = 'ASC'
        where_conditions.append('(id >= %s)')
        where_args.append(int(previous_page_last_batch_id))
    elif next_page_first_batch_id is not None:
        order_by_direction = 'DESC'
        where_conditions.append('(id <= %s)')
        where_args.append(int(next_page_first_batch_id))
    else:
        order_by_direction = 'DESC'
```

Our existing SQL query changes to have `LIMIT 51` instead of `LIMIT 50` and `DESC` becomes
`order_by_direction`.

We also need to add that new SQL query and modify what we return:

```python3
if previous_page_last_batch_id is not None:
    next_row_after_last_shown = await db.select_and_fetchone(
        f'select id from batches where id < {previous_page_last_batch_id} limit 1')

    has_next_page = next_row_after_last_shown is not None
    if has_next_page:
        next_page_first_batch_id = next_row_after_last_shown['id']
    else:
        next_page_first_batch_id = None

    has_previous_page = len(batches) == 51
    if has_previous_page:
        previous_page_last_batch_id = batches[-1]['id']
    else:
        previous_page_last_batch_id = None
elif next_page_first_batch_id is not None:
    previous_row_before_first_shown = await db.select_and_fetchone(
        f'select id from batches where id > {next_page_first_batch_id} limit 1')

    has_previous_page = previous_row_before_first_shown is not None
    if has_previous_page:
        previous_page_last_batch_id = previous_row_before_first_shown['id']
    else:
        previous_page_last_batch_id = None

    has_next_page = len(batches) == 51
    if has_next_page:
        next_page_first_batch_id = batches[-1]['id']
    else:
        next_page_first_batch_id = None
else:
    previous_page_last_batch_id = None  # the user went to /batches, so we're showing them the most recent batches

    has_next_page = len(batches) == 51
    if has_next_page:
        next_page_first_batch_id = batches[-1]['id']
    else:
        next_page_first_batch_id = None

batches = batches[:50]  # return at most 50 batches
return (batches, previous_page_last_batch_id, next_page_first_batch_id)
```

Wow, what a mouthful! UIs are hard!

Finally, the UI will have to change slightly. The page_context should remove `last_batch_id` and
instead have `previous_page_last_batch_id` and `next_page_first_batch_id`. The HTML will be very similar to
what you're currently doing:

```html
<div class="next-page-buttons">
  {% if previous_page_last_batch_id is not none %}
  <form method="GET" action="{{ base_path }}/batches">
    {% if q is not none %}
    <input type="hidden" name="q" value="{{ q }}" />
    {% endif %}
    <input type="hidden" name="previous_page_last_batch_id" value="{{ previous_page_last_batch_id }}" />
    <button>
     Previous page
   </button>
  </form>
  {% endif %}

  {% if next_page_first_batch_id is not none %}
  <form method="GET" action="{{ base_path }}/batches">
    {% if q is not none %}
    <input type="hidden" name="q" value="{{ q }}" />
    {% endif %}
    <input type="hidden" name="next_page_first_batch_id" value="{{ next_page_first_batch_id }}" />
    <input type="hidden" name="direction" value="lesser" />
    <button>
      Next page
    </button>
  </form>
  {% endif %}
</div>
```

File: batch/batch/front_end/front_end.py [hunk 1]
```diff
@@ -504,34 +513,43 @@ async def _query_batches(request, user):
...
                in db.select_and_fetchall(sql, sql_args)]
+    if sort_order == 'ASC':
+        batches.reverse()
+    if True:
+        first_batch_id = batches[0]['id']
+    else:
+        first_batch_id = None
 
-    if len(batches) == 50:
+    if len(batches) == 20:
```
[the remark this answers is not in the document]
Comment (CDiaz96) [reply batch/batch/front_end/front_end.py:531]: The limit should stay at 50. We changed it to 20 for the purpose of having multiple pages of batches created. 

oh okay. So when we do `LIMIT 51` it takes into account the 50 batches that were created and in essence uses that additional length to recognize there aren't any more batches, therefor the next page button would not be displayed.  I think I'm understanding that correctly.

**Comment by johnc1231** [comment]
@CDiaz96  and @danking, is this still relevant?

**Comment by danking** [comment]
We'll close until we have time to come back to this.

**Pull request closed without merging**